src/Entity/SmsMessage.php

```diff
@@ -587,12 +594,92 @@ public static function convertFromSmsMessage(StdSmsMessageInterface $sms_message
    */
   public static function postDelete(EntityStorageInterface $storage, array $entities) {
     parent::postDelete($storage, $entities);
+    $results = [];
+    $reports = [];
+    /** @var \Drupal\sms\Entity\SmsMessageInterface $sms_message */
     foreach ($entities as $sms_message) {
-      $sms_message->getResult() && $sms_message->getResult()->delete();
+      // Since the $sms_message can have both in-memory and stored objects, only
+      // need to delete actual stored entities.
+      if (($result = $sms_message->getResult()) && $result instanceof EntityInterface) {
+        $results[] = $result;
+      }
       foreach ($sms_message->getReports() as $report) {
-        $report->delete();
+        if ($report instanceof EntityInterface) {
+          $reports[] = $report;
+        }
       }
     }
+    \Drupal::entityTypeManager()->getStorage('sms_result')->delete($results);
+    \Drupal::entityTypeManager()->getStorage('sms_report')->delete($reports);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function postSave(EntityStorageInterface $storage, $update = TRUE) {
+    parent::postSave($storage, $update);
+    // Save the result and reports in the static cache.
+    if ($this->result) {
+      $result_entity = $this->convertFromMessageResult($this->result);
+      $result_entity
+        ->setSmsMessage($this)
+        ->save();
+
+      foreach ($this->result->getReports() as $report) {
+        $report_entity = $this->convertFromDeliveryReport($report);
+        $report_entity
+          ->setSmsMessage($this)
+          ->save();
+      }
+      // Unset $this->result as we don't need it anymore after save.
+      unset($this->result);
+    }
+  }
+
+  /**
+   * Converts a plain SMS message result into an SMS message result entity.
+   *
+   * @param \Drupal\sms\Message\SmsMessageResultInterface $sms_result
+   *   A plain SMS message result.
+   *
+   * @return \Drupal\sms\Entity\SmsMessageResultInterface
+   *   An SMS message result entity that can be saved.
+   */
+  protected function convertFromMessageResult(StdMessageResultInterface $sms_result) {
```

Comment:
TBH, I can't remember anymore. Again, I think it was an intermediate version (prior to the temporary $this->result storage), where I needed to save the result before returning.
And I didn't want to expose an API where converting to an entity automatically saves it.

This I can revert now.